spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DatastoreEntityConverterImpl.java

```
@@ -0,0 +1,130 @@
+	public <R> R read(Class<R> aClass, Entity entity) {
+		DatastorePersistentEntity<R> persistentEntity = (DatastorePersistentEntity<R>) mappingContext
+				.getPersistentEntity(aClass);
+		PreferredConstructor<?, DatastorePersistentProperty> persistenceConstructor = persistentEntity
+				.getPersistenceConstructor();
+
+		EntityPropertyValueProvider propertyValueProvider = new EntityPropertyValueProvider(entity);
+
+		ParameterValueProvider<DatastorePersistentProperty> parameterValueProvider =
+				new PersistentEntityParameterValueProvider<>(persistentEntity, propertyValueProvider, null);
+
+		EntityInstantiator instantiator = this.instantiators.getInstantiatorFor(persistentEntity);
+		R instance = instantiator.createInstance(persistentEntity, parameterValueProvider);
+		PersistentPropertyAccessor accessor = persistentEntity.getPropertyAccessor(instance);
+
+		persistentEntity.doWithProperties(
+				(PropertyHandler<DatastorePersistentProperty>) datastorePersistentProperty -> {
+					Object value = propertyValueProvider.getPropertyValue(datastorePersistentProperty);
+					accessor.setProperty(datastorePersistentProperty, value);
+				});
+
+		return instance;
+	}
+
+	@Override
+	public void write(Object source, Entity.Builder sink) {
+		DatastorePersistentEntity<?> persistentEntity = this.mappingContext
+				.getPersistentEntity(source.getClass());
+		PersistentPropertyAccessor accessor = persistentEntity
+				.getPropertyAccessor(source);
+		persistentEntity.doWithProperties(
+				(PropertyHandler<DatastorePersistentProperty>) datastorePersistentProperty -> {
+					writeProperty(sink, accessor, datastorePersistentProperty);
+				});
+	}
+
+	private <T> void writeProperty(Entity.Builder sink, PersistentPropertyAccessor accessor,
+			DatastorePersistentProperty persistentProperty) {
+
+		Object propertyVal = accessor.getProperty(persistentProperty);
+		Function<T, Value<?>> valueFactory = (Function<T, Value<?>>) valFactories.get(persistentProperty.getType());
```

This function might not be found if the type isn't supported, and would result in a npe on the next line. I think you should check for null here and throw an exception with a message about unsupported type or return false to signal that the write failed.  I think you should also add a unit test that tests this case where a property is an unsupported type.

---

good point